memq-client/src/main/java/com/pinterest/memq/client/commons2/network/NetworkClient.java

```diff
@@ -177,32 +181,42 @@ public CompletableFuture<ResponsePacket> send(InetSocketAddress socketAddress, R
     return returnFuture;
   }
 
+  /**
+   * Acquire channel for the given address by checking if it exists in the channelPool.
+   * If not, create it and register it in the channelPool for future use.
+   *
+   * @param socketAddress
+   * @return the ChannelFuture for this address
+   * @throws ExecutionException
+   * @throws InterruptedException
+   */
   protected ChannelFuture acquireChannel(InetSocketAddress socketAddress) throws ExecutionException, InterruptedException {
-    if (isChannelUnavailable(socketAddress)) {
-      synchronized (this) {
-        if (isChannelUnavailable(socketAddress)) {
-          if (connectFuture != null && connectFuture.channel() != null) {
-            // destination address is different from current connection's remote address
-            connectFuture.channel().close().await();
-          }
+    ChannelFuture existing = channelPool.get(socketAddress);
+    if (existing == null || !existing.channel().isActive()) {
+      synchronized (getPoolLock(socketAddress)) {
```

Comment:
It seems at least in the tests, we get here through these calls:

```
    Future<ResponsePacket> f1 = client.send(InetSocketAddress.createUnresolved(LOCALHOST_STRING, port), r1);
    Future<ResponsePacket> f2 = client.send(InetSocketAddress.createUnresolved(LOCALHOST_STRING, port), r2);
    Future<ResponsePacket> f3 = client.send(InetSocketAddress.createUnresolved(LOCALHOST_STRING, port), r3);
```

Each sending a different socket address, so synchronization at line 196 wouldn't prevent simultaneous access. Is that the intended behavior?

Reply:
yes, there is no need for blocking simultaneous access across different socket addresses. Channels with different socket addresses can be created simultaneously without side-effects, but we mainly want to prevent a situation where the same socket address channel is being created simultaneously and redundantly.